src/toolkits/style_transfer/style_transfer.cpp

```diff
@@ -0,0 +1,269 @@
+  auto out_end = out_it + h * w * c;
+  while (out_it != out_end) {
+    *out_it = *src_ptr / 255.f;
+    ++src_ptr;
+    ++out_it;
+  }
+
+  return shared_float_array::wrap(std::move(result_array), {h, w, c});
+}
+
+flex_int estimate_max_iterations(flex_int num_styles, flex_int batch_size) {
+  return (int)(num_styles * 10000.0f / batch_size);
+}
+
+}  // namespace
+
+void style_transfer::init_options(
+    const std::map<std::string, flexible_type>& opts) {
+  options.create_integer_option(
+      "batch_size",
+      "The number of images to process for each training iteration",
+      FLEX_UNDEFINED, 1, std::numeric_limits<int>::max());
+
+  options.create_integer_option(
+      "max_iterations",
+      "Maximum number of iterations to perform during training", FLEX_UNDEFINED,
+      1, std::numeric_limits<int>::max());
+
+  options.create_integer_option(
+      "random_seed",
+      "Seed for random weight initialization and sampling during training",
+      FLEX_UNDEFINED, std::numeric_limits<int>::min(),
+      std::numeric_limits<int>::max());
+
+  options.create_integer_option(
+      "num_styles", "The number of styles present in the model", FLEX_UNDEFINED,
+      1, std::numeric_limits<int>::max());
+
+  options.set_options(opts);
+
+  add_or_update_state(flexmap_to_varmap(options.current_option_values()));
+}
+
+size_t style_transfer::get_version() const { return STYLE_TRANSFER_VERSION; }
+
+void style_transfer::save_impl(oarchive& oarc) const {
+  variant_deep_save(state, oarc);
+  oarc << m_resnet_spec->export_params_view();
+}
+
+void style_transfer::load_version(iarchive& iarc, size_t version) {
+  variant_deep_load(state, iarc);
+
+  float_array_map nn_params;
+  iarc >> nn_params;
+
+  m_resnet_spec =
+      init_resnet(variant_get_value<size_t>(state.at("num_styles")));
+  m_resnet_spec->update_params(nn_params);
+}
+
+std::unique_ptr<compute_context> style_transfer::create_compute_context()
+    const {
+  return compute_context::create();
+}
+
+std::unique_ptr<data_iterator> style_transfer::create_iterator(
+    gl_sarray style, gl_sarray content, bool repeat, int random_seed) const {
+  data_iterator::parameters iterator_params;
+
+  iterator_params.style = std::move(style);
+  iterator_params.content = std::move(content);
+  iterator_params.repeat = repeat;
+  iterator_params.random_seed = random_seed;
+
+  return create_iterator(iterator_params);
+}
+
+std::unique_ptr<data_iterator> style_transfer::create_iterator(
+    data_iterator::parameters iterator_params) const {
+  return std::unique_ptr<data_iterator>(
+      new style_transfer_data_iterator(iterator_params));
+}
+
+void style_transfer::infer_derived_options() {
+  if (read_state<flexible_type>("batch_size") == FLEX_UNDEFINED) {
+    flex_int batch_size = DEFAULT_BATCH_SIZE;
+
+    size_t memory_budget = m_training_compute_context->memory_budget();
+    if (memory_budget < MEMORY_REQUIRED_FOR_DEFAULT_BATCH_SIZE) {
+      batch_size /= 2;
+    }
```

Comment:
Is this copied from the Python ST or from the C++ OD? Do we know this is actually the right threshold to use?

Reply:
Copied this from C++ OD. Not sure if it's the right threshold to use.

Reply:
If the current MPS implementation we're shipping in 5.8 doesn't adjust batch size due to memory, let's not do it here, either